calculator/calculator.py

```diff
@@ -1,15 +1,57 @@
 import re
-
-
-class Calculator:
-
-    @classmethod
-    def calculate(cls, expression=str):
-        if type(expression) is not str:
-            raise TypeError("Expression input should be of str type.")
-
-        # TODO
-
-
-
-
+import math
+from typing import Union
+from statistics import mean
+
+
+_NUMBER_REGEX = r'\d+(?:\.\d+)?'
+_NUMBER_NEGATIVE_OPT_REGEX = f'-?{_NUMBER_REGEX}'
+_LEFT_PARENTHESIS_REGEX = r'(?:\()'
+_RIGHT_PARENTHESIS_REGEX = r'(?:\))'
+
+
+def calculate(expression: str) -> Union[int, float]:
+    if type(expression) is not str:
+        raise TypeError("Expression input should be of str type.")
+
+    return _calculate(expression.replace(" ", ""))
+
+def _calculate(expression: str) -> Union[int, float]:
+    if _is_end_of_calculation(expression):
+        return expression
+
+    for regex, handler in _TOKEN_SPECIFICATION.items():
+        match = re.search(regex, expression)
+        if match:
+            solved_step_expression = re.sub(match.re, str(handler(*match.groups())), expression, 1)
+            return _calculate(solved_step_expression)
+
+    raise ValueError(f'Expression: {expression} is not legal!')
+
+def _add_operation_variables(operation: str, is_single_variable: bool = False, is_negative_legal: bool = False) -> str:
+    number_regex = _NUMBER_NEGATIVE_OPT_REGEX if is_negative_legal else _NUMBER_REGEX
+    return f'(?P<NUMBER>{number_regex}){operation}' if is_single_variable else f'(?P<FIRST_NUMBER>{number_regex})' \
+                                                                               f'{operation}' \
+                                                                               f'(?P<SECOND_NUMBER>{number_regex})'
+
+def _is_end_of_calculation(expression: str) -> bool:
+    return re.fullmatch(f'{_NUMBER_NEGATIVE_OPT_REGEX}', expression) is not None
+
+
+_TOKEN_SPECIFICATION = {
+    f'{_LEFT_PARENTHESIS_REGEX}(?P<EXTRACTED_EXPRESSION>[^{_LEFT_PARENTHESIS_REGEX}{_RIGHT_PARENTHESIS_REGEX}]*)'
+    f'{_RIGHT_PARENTHESIS_REGEX}': lambda x: _calculate(x),
+    _add_operation_variables(r'(?:!)', True, False): lambda x: math.factorial(float(x)),
+    _add_operation_variables(r'(?:~)', True, True): lambda x: -1 * float(x),
+    _add_operation_variables(r'(?:@)', False, True): lambda x, y: mean([float(x), float(y)]),
+    _add_operation_variables(r'(?:\&)', False, True): lambda x, y: min(float(x), float(y)),
+    _add_operation_variables(r'(?:\$)', False, True): lambda x, y: max(float(x), float(y)),
+    _add_operation_variables(r'(?:\^)', False, True): lambda x, y: float(x) ** float(y),
+    _add_operation_variables(r'(?:\*)', False, True): lambda x, y: float(x) * float(y),
+    _add_operation_variables(r'(?:/)', False, True): lambda x, y: float(x) / float(y),
+    _add_operation_variables(r'(?:\+)', False, True): lambda x, y: float(x) + float(y),
+    _add_operation_variables(r'(?:-)', False, True): lambda x, y: float(x) - float(y),
```

Comment:
It will be more clean if the float casting of the numbers string will be in the function that calling to the handler.
You have casting duplication in each function, if you decide that you work only with floats, cast it with list comprehension/map in `_calculate`.

Reply:
operands may be a string, int or tuple, therefor casting in _calculate might be adding code that won't be much more cleaner.  Also because of the factorial now i use int also in its lambda instead of float, so i think leaving it as it is would be a better idea. what do you think?

Reply:
It's calculator - you work only with numbers (floats as private case) so operands cannot be string, tuple, etc. 
There are a lot of operations here that don't require floats but you cast the operands to float - it's seems that float is your default, so I think you should cast it in `_calculate` and here perform pure operations. If you have specific operation that use `int`, cast it to int from float, no big deal.
Also, It not really matters except if you will add new operation tommorow you will have cast the operands again and it looks like black work.
This is ahestetics, no something important.